sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveReferencesInAggregate.scala

```diff
@@ -0,0 +1,183 @@
+ * rule `ResolveReferences`. The column resolution order for [[Aggregate]] is:
+ * 1. Resolves the column to [[AttributeReference]] with the output of the child plan. This
+ *    includes metadata columns as well.
+ * 2. Resolves the column to a literal function which is allowed to be invoked without braces, e.g.
+ *    `SELECT col, current_date FROM t`.
+ * 3. Resolves the column to [[LateralColumnAliasReference]] if it references the alias defined
+ *    previously in the SELECT list. The rule `ResolveLateralColumnAliasReference` will further
+ *    resolve [[LateralColumnAliasReference]] and rewrite the plan. This is to support SQL pattern
+ *    like `SELECT col1 + 1 AS x, x + 1 AS y, y + 1 AS z FROM t`.
+ * 4. If the group by expressions contain an unresolved column whose name matches an alias in the
+ *    SELECT list, resolves that unresolved column to the alias if the alias is resolved. This is to
+ *    support SQL pattern like `SELECT a + b AS c, max(col) FROM t GROUP BY c`.
+ * 5. If the group by expressions only have one single unresolved column named ALL, expanded it to
+ *    include all non-aggregate columns in the SELECT list. This is to support SQL pattern like
+ *    `SELECT col1, col2, agg_expr(...) FROM t GROUP BY ALL`.
+ * 6. Resolves the column to outer references with the outer plan if we are resolving subquery
+ *    expressions.
+ */
+object ResolveReferencesInAggregate extends SQLConfHelper with ColumnResolutionHelper {
+  def apply(a: Aggregate): Aggregate = {
+    val planForResolve = a.child match {
+      // SPARK-25942: Resolves aggregate expressions with `AppendColumns`'s children, instead of
+      // `AppendColumns`, because `AppendColumns`'s serializer might produce conflict attribute
+      // names leading to ambiguous references exception.
+      case appendColumns: AppendColumns => appendColumns
+      case _ => a
+    }
+
+    val resolvedAggExprsNoOuter = a.aggregateExpressions
+      .map(resolveExpressionByPlanChildren(_, planForResolve, allowOuter = false)
+        .asInstanceOf[NamedExpression])
+
+    val resolvedGroupingExprs = a.groupingExpressions
+      .map(resolveExpressionByPlanChildren(_, planForResolve, allowOuter = false))
+      // SPARK-31670: Resolve Struct field in groupByExpressions and aggregateExpressions
+      // with CUBE/ROLLUP will be wrapped with alias like Alias(GetStructField, name) with
+      // different ExprId. This cause aggregateExpressions can't be replaced by expanded
+      // groupByExpressions in `ResolveGroupingAnalytics.constructAggregateExprs()`, we trim
+      // unnecessary alias of GetStructField here.
+      .map(trimTopLevelGetStructFieldAlias)
+
+    if (resolvedAggExprsNoOuter.forall(_.resolved)) {
+      val finalGroupExprs = resolveGroupByAll(
+        resolvedAggExprsNoOuter,
+        resolveGroupByAlias(resolvedAggExprsNoOuter, resolvedGroupingExprs)
+      ).map(resolveOuterRef)
```

Comment:
This is the actual change. Now we resolve group by alias and group by all before resolving outer references.